linux_os/guide/system/software/sudo/sudoers_no_command_negation/oval/shared.xml

```diff
@@ -0,0 +1,26 @@
+<def-group>
+     <definition class="compliance" id="{{{ rule_id }}}" version="1">
+     {{{ oval_metadata("Check that sudoers doesn't contain command negations") }}}
+     <criteria operator="AND">
+	     <criterion comment="Make sure that no command in user spec contains negation" test_ref="test_{{{ rule_id }}}" />
+     </criteria>
+  </definition>
+
+  <ind:textfilecontent54_test check="all" check_existence="none_exist"
+  comment="Make sure that no user spec in sudoers has a runas spec that includes root or ALL"
+	  id="test_{{{ rule_id }}}" version="1">
+  <ind:object object_ref="object_{{{ rule_id }}}" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_{{{ rule_id }}}" version="1">
+    <ind:filepath operation="pattern match">^/etc/sudoers(\.d/.*)?$</ind:filepath>
+    <!-- The regex idea: <user list> <host list> = (<the whole command without negation>,)* <command with negation> <whatever>
+         where a command is <runas spec>?<anything except , or !>+,
+           - ',' is a command delimiter, while
+           - '!' outside of a runas spec is a command negation we are after,
+         The capturing group holds the offending command.
+    -->
+    <ind:pattern operation="pattern match">^(?:\s*[^#][^=]+)=(?:\s*(\([^\)]+\))?\s*(?!\s*\()[^,!\n]+,)*\s*(?:\([^\)]+\))?\s*(?!\s*\()([^,\n]*!\S+).*</ind:pattern>
```

Comment:
```suggestion
    <ind:pattern operation="pattern match">^(?:\s*[^#][^=]+)=(?:\s*(\(?:[^\)]+\))?\s*(?!\s*\()[^,!\n]+,)*\s*(?:\([^\)]+\))?\s*(?!\s*\()([^,\n]*!\S+).*</ind:pattern>
```
This avoids capturing runas specs with a `!`.

Reply:
I was hoping that the captured spec would somehow pop up in OVAL results, but that was never the case, so I will do it like that.

Reply:
Hmm, I know that it emerges in the OpenSCAP DEVEL logs, in the collected item as `subexpression`.
It should show up in  the OVAL results too.

Reply:
It should avoid capturing the runas spec, like `(bob !alice)`.
And actually, the suggestion is above.
Here is correct one:
```suggestion
    <ind:pattern operation="pattern match">^(?:\s*[^#][^=]+)=(?:\s*(?:\([^\)]+\))?\s*(?!\s*\()[^,!\n]+,)*\s*(?:\([^\)]+\))?\s*(?!\s*\()([^,\n]*!\S+).*</ind:pattern>
```